Plugin Manager - Listing by status

piloujp opened the pull request:
List plugins in different tables depending on their status.
<img width="1896" height="815" alt="Capture d'écran 2025-09-21 183424" src="https://github.com/user-attachments/assets/e221a63b-7c94-44d3-b74e-b275ec70a090" />

piloujp commented:
I added a commit to display installed plugins first and not installed ones last.
<img width="1895" height="811" alt="Capture d'écran 2025-09-21 221746" src="https://github.com/user-attachments/assets/336edb84-3e04-4fe4-bb4c-0a23a0043ba3" />

piloujp commented:
There is one point I am not completely satisfied with in this PR. When installing/uninstalling or enabling/disabling a plugin, normally this plugin line stays the active one. Problem is, when doing one of these operations, the plugin is displayed in another table which could be on another page too. In this case, the plugin details box on the right is messed up, and an error log is created. To avoid that, I had to remove that functionality and the first line of the page becomes the active one.
Perhaps someone has an idea how to maintain this line activation without causing an error.
The code modified is at the end of the functions `processActionDo_SomeAction_` in file `../includes/classes/ViewBuilders/PluginManagerController.php` where I removed ` . '&' . $this->colKeylink()` from the link.

piloujp commented:
> What was the reasoning for changing the meaning of a plugin's `status` values?

To change the order of display and get the table with enabled plugins first instead of the one with plugins not installed. Sorting by status is always in this order: 0->1->2
It is possible to do a custom sorting with MySQL like 1->2->0, but unfortunately Laravel is taking care of the query...

piloujp commented:
New commit to remove pagination and solving line activation problem.

lat9 commented:
> I added a commit to display installed plugins first and not installed ones last. <img alt="Capture d&apos;écran 2025-09-21 221746" width="1895" height="811" src="https://private-user-images.githubusercontent.com/117885587/492026289-336edb84-3e04-4fe4-bb4c-0a23a0043ba3.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg1Nzc2ODcsIm5iZiI6MTc1ODU3NzM4NywicGF0aCI6Ii8xMTc4ODU1ODcvNDkyMDI2Mjg5LTMzNmVkYjg0LTNlMDQtNGZlNC1iYjRjLTBhMjNhMDA0M2JhMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkyMlQyMTQzMDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNmRkNTNiZGE3NjZmZDY2NWRiNDcxM2ZmZjE0MTRmNjljNzVjZGZkYzRiMGRjNGNkYTk5Zjg0MTg0Y2E2ODhjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9._NcLvWv-deFXd48IZEkbfwpyhtmCEmrdm8hOPDr03Z8">

These columns need to line up.

drbyte commented:
To get the columns to line up when I did similarly in the Modules page, I applied these styles to the various TH column header cells:
```css
      <style>
          .w-40 {width: 40%}
          .w-20 {width: 20%}
          .w-10 {width: 10%}
      </style>
```

piloujp commented:
OK, I moved template code in `plugin_manager.php` file and put back `table_view.php` to its original state. Everything is working fine.
It is probably better like this, the template needs rework to be used in different pages, as the `$formater`. Like this, it can be done without interfering with the plugin manager page.

piloujp commented:
I found a missing `</div>` and realized it was actually missing in `table_view.php` too. Another PR to do.

piloujp commented:
What about the code for the filter that is no implemented. It is still there, now from line 123 to 136 in `plugin_manager.php`?
As plugins are displayed by status in different tables, this filter, if activated, will only allow to display one table/category at a time.
Perhaps it could be removed from here? Anyway, it is still in `table_view.php`.

drbyte commented:
Oy. I meant `/admin/includes/templates/plugin_manager.php`

drbyte commented:
And, to clarify, was the renumbering of `status` because you're relying on sql queries to order things?

Could the status values not be left as they were, and then display each of the tables according to an array of statuses?

Maybe:
```diff
-            for ($i = 0; $i < 3; $i++) {
+            foreach ([2, 0, 1] as $i) {
```

drbyte commented:
Given that the filter feature is not implemented in this plugin_manager page, I have no objection to removing that from /admin/includes/templates/plugin_manager.php template. Could remove the paginator from that file too, since not used there.
And then table_view.php can still keep those stubs for future use.

piloujp commented:
> Oy. I meant `/admin/includes/templates/plugin_manager.php` as the page-specific variant of table_view.php

Oups, misunderstanding...

piloujp commented:
After applying the sorting with the `foreach` loop, I reverted all status numbers changes.
Well, things got a little bit messy, but at the end the solution is simpler.
There is only one file modified file, `admin/plugin_manager.php` and a new file `admin/includes/templates/plugin_manager.php`, which is a custom version of `table_view.php`.
Although I reverted the testFramework change, it seems now it is failing somewhere else, probably because of the new file.

piloujp commented:
Ahrg, now the default selected line when opening the page becomes the first plugin in alphabetic order of its plugin key, which not necessary in the installed plugins table...
It is defined by query's order by value in `includes/classes/ViewBuilders/PluginManagerDataSource.php`:
`return (new PluginControl())->query()->orderBy('status', 'asc')->orderBy('name')->orderBy('unique_key');`
If I put back status in here, the default active line will be the first one in `Not Installed` plugins table... Which means to fix that, the status order must be `0->Installed(enabled)`, `1->Installed(disabled)`, `3->Not Installed` ...

drbyte commented:
I think the default-selected line should be determined separately from the query that retrieves all the records, not by an orderBy clause.

piloujp commented:
> I think the default-selected line should be determined separately from the query that retrieves all the records, not by an orderBy clause.

After a few hours working on this, I think it is impossible without rewriting the `SimpleDataFormatter` class. This there that the first row of the dataset is used by default in method `isRowSelected()` which calls `currentRowFromRequest()` where, on line 77, the code that make this choice is written.
These methods are called three times in the template that built the page. I was able to write some code that do the same functionality directly in the template for two of them but not for the third one which happens when the plugin details box on the right is built. This last code is so much nested in the `box` and `boxTableBlock` classes that again some class rewrite would be necessary.
In actual ZC, just modifying `SimpleDataFormatter.php` by replacing line 77
`$result = $this->resultSet->getCollection()->first();`
with this:
`$result = $this->resultSet->getCollection()->where('status', 1)->first();`
will do the trick, but this hard codded change will certainly be a problem if this method is to be used on other pages construct in the futur.

drbyte commented:
Thanks for exploring that.

Today I had some time to rollback this branch to commit https://github.com/zencart/zencart/pull/7303/commits/ad838454548e047c69e9b0dd51f358f890eeedc7 ("New plugin _manager.php template file and simplified sorting. Reverted back all status number changes").
I made no database changes.

And I found that the UI worked as-expected (proper line-selection, proper sidebox operation) for everything.

What were you seeing as the problem at that point? (other than I saw that the test suite had an issue, which I'm investigating)

piloujp commented:
> the default selected line when opening the page becomes the first plugin in alphabetic order of its plugin key, which is not necessary in the installed plugins table...

If you have the first plugin in alphabetic order installed (like AbuseIPDB), it looks OK. But if this plugin is disabled or not installed, it will still be the active one by default, although it will be displayed in the second or third table. In my test with AbuseIPDB not installed, but more than 16 other plugins installed, it was not even visible without scrolling, but still the default active line.

drbyte commented:
Try this:
```diff
class PluginManagerDataSource extends DataTableDataSource
{
    protected function buildInitialQuery(): Builder
    {
-        return (new PluginControl())->query()->orderBy('name')->orderBy('unique_key');
+        $statusSort = [
+            1, // enabled
+            2, // disabled
+            0, // not installed
+        ];
+        return PluginControl::query()
+            ->orderByRaw(
+                "FIELD(status, " . implode(',', $statusSort) . ")"
+            )
+            ->orderBy('name')
+            ->orderBy('unique_key');
    }
}
```

piloujp commented:
Excellent! How did you found that? One of the first options I explored to solve this sorting problem led me there, in class `PluginControl` that actually uses Laravel where I gave up. If I had known of this `orderByRaw` method...
Now it looks good, all performing as expected with a minimum code modification.
